perf(transformer/styled-components): remove quasis and expressions in batch

Dunqing opened the pull request:
Remove `quasis` and `expressions` in batch instead of removing them one by one, so that it would be more efficient. Considering that the expression's removal is unlikely to be common, I add a `has_deletion_marker` flag to avoid unnecessary `retain` calls, so that we can keep the performance as before in this situation.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12256?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#12256** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12256?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12256?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12240** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12240?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12224** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12224?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12256).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel reviewed (commented):
As you say, this case is uncommon, but yes using `retain` is ideal to avoid poor perf in odd cases where there are tons of quasis and many get deleted. 👍

If we do #12213, then probably we'll be able to make the main loop use `retain`, and then we won't need the extra loops to remove nodes at the end.

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -911,11 +914,10 @@ fn minify_template_literal<'a>(lit: &mut TemplateLiteral<'a>, ast: AstBuilder<'a
-                quasis.remove(quasi_index);
-                expressions.remove(quasi_index - 1);
+                // Mark previous quasi for deletion, it will be deleted all of quasis are processed.
+                quasis[quasi_index - 1].value.raw = QUASI_DELETION_MARKER;
```
It's *this* quasi which should be deleted, not the previous one.

overlookmotel replied:
Weird that all tests still passing with this mistake. Don't the tests cover this? Or am I mistaken?

overlookmotel replied:
I've got it now. I'll update the comments in a follow-on PR to reflect this change in the logic.

overlookmotel commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -911,11 +914,10 @@ fn minify_template_literal<'a>(lit: &mut TemplateLiteral<'a>, ast: AstBuilder<'a
                 quasis[quasi_index - 1].span.end = quasi.span.end;
-
-                // Remove this quasi and the preceding expression.
-                // TODO: Remove scopes, symbols, and references for removed `Expression`.
```
This comment is still relevant. Don't delete me!

Dunqing commented on crates/oxc_transformer/src/plugins/styled_components.rs:
```diff
@@ -911,11 +914,10 @@ fn minify_template_literal<'a>(lit: &mut TemplateLiteral<'a>, ast: AstBuilder<'a
-                // TODO: Remove scopes, symbols, and references for removed `Expression`.
-                quasis.remove(quasi_index);
-                expressions.remove(quasi_index - 1);
+                // Mark previous quasi for deletion, it will be deleted all of quasis are processed.
+                quasis[quasi_index - 1].value.raw = QUASI_DELETION_MARKER;
+                has_deletion_marker = true;
+                quasi_index += 1;
```
Sounds good, I always follow the best performance way!

overlookmotel replied:
OK cool. Probably a good idea to define a const for the sentinel span so it's clear what's going on.

```rs
/// `Span` used as a sentinel indicating quasi should be removed.
/// Source text is limited to max `u32::MAX` bytes, so it's impossible for a `TemplateElement`
/// to have this span, because it's always followed by (at minimum) a '`'.
const REMOVE_SENTINEL: Span = Span::new(u32::MAX, u32::MAX);
``

overlookmotel commented:
### Merge activity

* **Jul 15, 9:55 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
perf(transformer/styled-components): remove quasis and expressions in batch (#12256)

Remove `quasis` and `expressions` in batch instead of removing them one by one, so that it would be more efficient. Considering that the expression's removal is unlikely to be common, I add a `has_deletion_marker` flag to avoid unnecessary `retain` calls, so that we can keep the performance as before in this situation.
